engine/include/cubos/engine/ui/canvas/draw_list.hpp

```diff
@@ -0,0 +1,158 @@
+                : type(commandType)
+            {
+            }
+
+            /// @brief Constructs.
+            /// @param commandType Command type.
+            /// @param drawCommand UI draw command.
+            /// @param dataOffset Position of the command data in the draw list.
+            UIDrawListEntry(const UIDrawType& commandType, const UIDrawCommand& drawCommand, size_t dataOffset)
+                : type(commandType)
+                , command(drawCommand)
+                , offset(dataOffset)
+            {
+            }
+
+            const UIDrawType& type; ///< Command type.
+            UIDrawCommand command;  ///< UI draw command.
+            size_t offset;          ///< Position of the command data in the draw list.
+        };
+
+        /// @brief Builds a UIDrawListEntry.
+        /// @ingroup ui-canvas-plugin
+        class UIDrawListEntryBuilder
+        {
+        public:
+            /// @brief Constructs builder with required data.
+            /// @param entry Reference to the associated entry.
+            UIDrawListEntryBuilder(UIDrawListEntry& entry);
+
+            /// @brief Adds a texture to the entry.
+            /// @param texIndex Index of the texture to add.
+            /// @param texture Texture to add.
+            /// @param sampler Sampler to be used with the texture.
+            /// @returns The builder.
+            UIDrawListEntryBuilder withTexture(int texIndex, cubos::core::gl::Texture2D texture,
+                                               cubos::core::gl::Sampler sampler = nullptr);
+
+            /// @brief Gets the built entry.
+            /// @return Built UIDrawListEntry.
+            UIDrawListEntry entry() const;
+
+        private:
+            UIDrawListEntry& mEntry;
+        };
+
+        /// @brief Constructs a draw list.
+        UIDrawList();
+
+        /// @brief Move constructs.
+        /// @param other Other draw list.
+        UIDrawList(UIDrawList&& other) noexcept;
+
+        /// @brief Copy constructs.
+        /// @param other Other draw list.
+        UIDrawList(const UIDrawList& other) noexcept;
+
+        ~UIDrawList();
+
+        /// @brief Adds entry to the draw list.
+        /// @param type Command type.
+        /// @param command UI draw command.
+        /// @param data Command data.
+        void push(const UIDrawType& type, const UIDrawCommand& command, const void* data);
+
+        /// @brief Creates a new entry in the draw list.
+        /// @param type Command type.
+        /// @param vertexArray Vertex array handle.
+        /// @param vertexOffset Index of the first vertex to be drawn.
+        /// @param vertexCount Number of vertexes to draw.
+        /// @param data Command data.
+        /// @return A builder for the created entry.
+        UIDrawListEntryBuilder draw(const UIDrawType& type, core::gl::VertexArray vertexArray, size_t vertexOffset,
+                                    size_t vertexCount, const void* data);
```

Comment:
```suggestion
        UIDrawListEntryBuilder push(const UIDrawType& type, core::gl::VertexArray vertexArray, size_t vertexOffset,
                                    size_t vertexCount, const void* data);
```

I think you can leave it as `push` here, since this should just be a wrapper around the function above.